backend/server/routers/ctf.py

```diff
@@ -0,0 +1,283 @@
+            term_sum = sum(map(get_code, term.keys()))
+            print(f"{y}T{i} sum: {term_sum}")
+            if not is_even(term_sum):
+                print("failed: ", term)
+                return False
+
+    return True
+
+# TODO
+def term_sums_odd(data: PlannerData) -> bool:
+    """
+    Check that the sum of the course codes in odd terms is odd
+    """
+    is_odd: Callable[[int], bool] = lambda x: x % 2 == 1
+    print("Checking odd")
+    for year in data.plan[::2]:
+        # Exclude summer term + even terms
+        for term in year[1::2]:
+            term_sum = sum(map(get_code, term.keys()))
+            if not is_odd(term_sum):
+                print("failed: ", term)
+                return False
+    return True
+ 
+def comp1511_marks(data: PlannerData) -> bool:
+    """
+    Ollie must achieve a mark of 100 in COMP1511 to keep his scholarship
+    """
+    for year in data.plan:
+        for term in year:
+            for course in term:
+                _, marks = term[course]  # type: ignore
+                if course == "COMP1511":
+                    return marks == 100
+
+    return False
+
+
+def gen_ed_sum(data: PlannerData) -> bool:
+    """
+    The sum of GENED course codes must not exceed 2200
+    """
+    return sum(map(get_code, gen_eds(all_courses(data)))) <= 2200
+
+
+def gen_ed_faculty(data: PlannerData) -> bool:
+    """
+    Gen-Eds must all be from different faculties
+    """
+    gen_eds_facs = list(map(get_faculty, gen_eds(all_courses(data))))
+    return len(gen_eds_facs) == len(set(gen_eds_facs))
+
+
+def same_code_diff_faculty(data: PlannerData) -> bool:
+    """
+    Must take two courses with the same code but, from different faculties
+    """
+    codes = list(map(get_code, all_courses(data)))
+    # Can't have duplicate of a course since it's a set
+    return len(codes) != len(set(codes))
+
+
+def math_limit(data: PlannerData) -> bool:
+    """
+    In your N-th year, you can only take N + 1 math courses
+    """
+    for i, year in enumerate(data.plan, 1):
+        num_math = len([
+            course
+            for term in year
+            for course in term
+            if course.startswith("MATH")
+        ])
+        if num_math > i + 1:
+            return False
+
+    return True
+
+def six_threes_limit(data: PlannerData) -> bool:
+    """
+    There can by at most 6 occurrences of the number 3 in the entire
+    planner
+    """
+    all_codes = "".join(str(get_code(course)) for course in all_courses(data))
+    return all_codes.count("3") <= 6
+
+def comp1531_third_year(data: PlannerData) -> bool:
+    """
+    COMP1531 must be taken in the third year
+    """
+    third_year = data.plan[2]
+    for term in third_year:
+        for course in term:
+            if course == "COMP1531":
+                return True
+
+    return False
+
+# (validator_func, message, Optional<flag>)
+requirements: list[tuple[Callable[[PlannerData], bool], str, Optional[str]]] = [
```

Comment:
thoughts on extracting the validator_func type to make this more readable?

Reply:
its already readable, skill diff